### PR TITLE
Add --format option to `borg diff`, resolve issue #4634

### PR DESCRIPTION
**Pull request by RF-Tar-Railt**
- [x] add `--format` in `diff_cmd.py`
- [x] rewrite `ItemDIff` in `item.pyx`
- [x] keep `--content-only` and `--json-lines`
- [x] rewrite part of `BaseFormatter` and subclass
- [x] write `DiffFormatter`
- [x] use `DiffFormatter` in `diff_cmd`  
- [x] test

Old PR with feedback: #7440

### CODE REVIEW

**Comment by ThomasWaldmann** [comment]
There still some unresponded feedback in the old PR feedback.

Also, now that you've freed your local master branch, you could do a `git rebase -i` onto an updated master (and maybe clean up the fix fix fix a bit), see the docs ticket describing this if unclear.

**Comment by RF-Tar-Railt** [comment]
Strange thing: This PR is not displayed in my Pull Requests interface, nor is it displayed in the Repo PRs when using gh mobile 🤔

**Comment by RF-Tar-Railt** [comment]
> Strange thing: This PR is not displayed in my Pull Requests interface, nor is it displayed in the Repo PRs when using gh mobile 🤔

restored after my comment. 🧐

File: src/borg/archive.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
 from getpass import getuser
 from io import BytesIO
 from itertools import groupby, zip_longest
+from typing import Iterator  # TODO: remove this if maintainers don't want it
```
Comment (ThomasWaldmann) [inline src/borg/archive.py:14]: we'll keep the type hints, but please remove all these related "TODOs".

i am generally happy with these, but they made this PR a bit larger than it would have been without them.

File: src/borg/testsuite/archiver/diff_cmd.py [hunk 2]
```diff
@@ -72,22 +72,20 @@ def test_basic_functionality(self):
         self.cmd(f"--repo={self.repository_location}", "create", "test1b", "input", "--chunker-params", "16,18,17,4095")
 
         def do_asserts(output, can_compare_ids, content_only=False):
-            # File contents changed (deleted and replaced with a new file)
-            change = "B" if can_compare_ids else "{:<19}".format("modified")
-            lines = output.splitlines()
+            lines: list = output.splitlines()
             assert "file_replaced" in output  # added to debug #3494
-            self.assert_line_exists(lines, f"{change}.*input/file_replaced")
+            self.assert_line_exists(lines, "modified.*input/file_replaced")
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/archiver/diff_cmd.py:77]: you removed the specific testing for the can_compare_ids True case:
it used to test for "B...filename" (B being the unit [bytes] of the actual size of the detected changes).

for the can_compare_ids False case, it can not compute the actual size of the changes, because the ids are computed differently and so it can't detect what's identical and what not.

File: src/borg/testsuite/archiver/diff_cmd.py [hunk 3]
```diff
@@ -96,16 +94,17 @@ def do_asserts(output, can_compare_ids, content_only=False):
...
+                    assert "input/link_replaced_by_file" in output
 
                 # Symlink target removed. Should not affect the symlink at all.
                 assert "input/link_target_removed" not in output
 
             # The inode has two links and the file contents changed. Borg
             # should notice the changes in both links. However, the symlink
             # pointing to the file is not changed.
-            change = "0 B" if can_compare_ids else "{:<19}".format("modified")
+            change = "modified.* B" if can_compare_ids else r"modified:  \(can't get size\)"
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/archiver/diff_cmd.py:107]: why did you remove the "0" from "change"?

File: src/borg/testsuite/archiver/diff_cmd.py [hunk 4]
```diff
@@ -294,6 +307,9 @@ def test_sort_option(self):
             "e_file_changed",
             "f_file_removed",
         ]
+        assert isinstance(output, str)
+        outputs = output.splitlines()
+        assert len(outputs) == len(expected)
         assert all(x in line for x, line in zip(expected, output.splitlines()))
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/archiver/diff_cmd.py:313]: you could use outputs here instead of calling splitlines again.

File: src/borg/helpers/parseformat.py [hunk 5]
```diff
@@ -963,6 +949,141 @@ def format_iso_time(self, key, item):
...
+        diff_data["change"] = " ".join([v for v in change if v])
+        diff_data["path"] = item.path
+        diff_data.update(self.item_data)
+        return diff_data
+
+    def format_other(self, key, diff: "ItemDiff"):
+        change = diff.changes().get(key)
+        if change is None:
+            return ""
+        return f"{change.diff_type}".ljust(27)
```
Comment (ThomasWaldmann) [inline src/borg/helpers/parseformat.py:1035]: add a comment about why 27.

Comment (ThomasWaldmann) [reply src/borg/helpers/parseformat.py:1035]: also, could be: `return ... if change is not None else ""`

File: src/borg/helpers/parseformat.py [hunk 6]
```diff
@@ -963,6 +949,141 @@ def format_iso_time(self, key, item):
...
+
+    def format_mode(self, diff: "ItemDiff", ft=False):
+        change = diff.type() if ft else diff.mode()
+        if change:
+            return f"[{change.diff_data['item1']} -> {change.diff_data['item2']}]"
+        return ""
+
+    def format_owner(self, diff: "ItemDiff", spec="owner"):
+        if spec != "owner":
+            change = diff.group() if spec == "group" else diff.user()
```
Comment (ThomasWaldmann) [inline src/borg/helpers/parseformat.py:1045]: please have a if / elif dispatcher for all valid spec values and raise ValueError in the else branch for unexpected values.

File: src/borg/helpers/parseformat.py [hunk 7]
```diff
@@ -963,6 +949,141 @@ def format_iso_time(self, key, item):
...
+                format_file_size(change.diff_data["added"], precision=1, sign=True),
+                format_file_size(-change.diff_data["removed"], precision=1, sign=True),
+            )
+        return ""
+
+    def format_time(self, key, diff: "ItemDiff"):
+        if key in diff.changes():
+            change = diff.changes()[key]
+            return f"[{key}: {change.diff_data['item1']} -> {change.diff_data['item2']}]"
+        return ""
```
Comment (ThomasWaldmann) [inline src/borg/helpers/parseformat.py:1078]: use `.get()` like you did in `format_other`?

File: src/borg/helpers/parseformat.py [hunk 8]
```diff
@@ -963,6 +949,141 @@ def format_iso_time(self, key, item):
...
+
+    def format_iso_time(self, key, diff: "ItemDiff"):
+        if key in diff.changes():
+            change = diff.changes()[key]
+            return f"[iso{key}: {change.diff_data['item1'].isoformat()} -> {change.diff_data['item2'].isoformat()}]"
+        return ""
```
Comment (ThomasWaldmann) [inline src/borg/helpers/parseformat.py:1084]: same here.

**Review by ThomasWaldmann — CHANGES_REQUESTED** [review]
parseformat review

**Comment by ThomasWaldmann** [comment]
@RF-Tar-Railt can you address the remaining feedback also?

**Comment by RF-Tar-Railt** [comment]
> k also?

yep. just that I have to do somethings first thesedays. I have drawn up some solutions.

File: src/borg/helpers/parseformat.py [hunk 9]
```diff
@@ -963,6 +949,141 @@ def format_iso_time(self, key, item):
...
+    def available_keys(cls):
+        from ..item import Item, ItemDiff
+
+        fake_item1 = Item(mode=0, path="", user="", group="", mtime=0, uid=0, gid=0)
+        fake_item2 = Item(mode=0, path="", user="", group="", mtime=0, uid=0, gid=0)
+        fake_diff = ItemDiff("", fake_item1, fake_item2, iter([]), iter([]))
+        formatter = cls("")
+        keys = []
+        keys.extend(formatter.call_keys.keys())
+        keys.extend(formatter.get_item_data(fake_diff).keys())
```
[the remark this answers is not in the document]
Comment (RF-Tar-Railt) [reply src/borg/helpers/parseformat.py:981]: As far as I know, "available_keys" seems to be only used in the help text.
(may even be enough to use `BaseFormatter.available_keys` 🤔)

**Review by ThomasWaldmann — COMMENTED** [review]
can we have a better commit comment here which described the ideas behind the changes?

renaming stuff and changing semantics in same commit is bad for review.

**Comment by RF-Tar-Railt** [comment]
If `available_keys()` is a nesserary design, I will revert commit [`602dea1`](https://github.com/borgbackup/borg/pull/7534/commits/602dea178108f94bcf0a0b6c5fdd3034a211988f)

**Comment by ThomasWaldmann** [comment]
> If available_keys() is a necessary design, I will revert commit 602dea1.

Guess the difference between the original implementation and your updated one is that the original looked at what keys we actually have in the data and made sure we have docs for **all** these keys.

Your implementation now does not look at the data and just makes sure we do not have more keys in the description dict than in the key groups. Guess it does not make sure that the generated docs describe all keys we have in the data.

**Comment by RF-Tar-Railt** [comment]
> > If available_keys() is a necessary design, I will revert commit 602dea1.
> 
> 
> 
> Guess the difference between the original implementation and your updated one is that the original looked at what keys we actually have in the data and made sure we have docs for **all** these keys.
> 
> 
> 
> Your implementation now does not look at the data and just makes sure we do not have more keys in the description dict than in the key groups. Guess it does not make sure that the generated docs describe all keys we have in the data.
> 
> 
> 
> 

So is it better to get all the keys from item according to the original implementation, or we set up a check function to ensure that possible Formatter in the future have written docs describe all keys we have in the data.?

File: src/borg/helpers/parseformat.py [hunk 10]
```diff
@@ -963,6 +915,125 @@ def format_iso_time(self, key, item):
...
+class DiffFormatter(BaseFormatter):
+    KEY_DESCRIPTIONS = {
+        "path": "archived file path",
+        "change": "all available changes",
+        "content": "file content change",
+        "mode": "file mode change",
+        "type": "file type change",
+        "owner": "file owner (user/group) change",
+        "user": "file user change",
+        "group": "file group change",
```
Comment (ThomasWaldmann) [inline src/borg/helpers/parseformat.py:927]: looks like some stuff is missing here (see KEY_GROUPS)?

Comment (RF-Tar-Railt) [reply src/borg/helpers/parseformat.py:927]: other Formatter also does?

Comment (ThomasWaldmann) [reply src/borg/helpers/parseformat.py:943]: ehrm, yeah, looks like. thanks for fixing these.

**Comment by ThomasWaldmann** [comment]
Guess I need to play a bit with the code before giving more feedback...

**Comment by ThomasWaldmann** [comment]
Some stuff i found while playing with it:
- help output is missing a blank before `(`:
```
  --format FORMAT       specify format for differences between archives(default: "{change} {path}{NL}")
```
- when using `isomtime` placeholder, it outputs `isomtime` but should be rather `mtime` (same for other iso...):
```
modified:  +9.0 kB  -9.0 kB [isomtime: 2023-06-06T11:58:38.614099+02:00 -> 2023-06-07T01:48:24.724694+02:00] setup.py
```
- when using `type` it shows a type change `[- -> -]` even if the file type did not change, but just other mode bits.
- `link` says `changed link`, but not how.
- `link`, `blkdev`, `chrdev`, `fifo`, `directory` these somehow feel wrong, maybe need more thinking about practical use.

**Comment by RF-Tar-Railt** [comment]
> Some stuff i found while playing with it:
> 
> * help output is missing a blank before `(`:
> 
> ```
>   --format FORMAT       specify format for differences between archives(default: "{change} {path}{NL}")
> ```
> 
> * when using `isomtime` placeholder, it outputs `isomtime` but should be rather `mtime` (same for other iso...):
> 
> ```
> modified:  +9.0 kB  -9.0 kB [isomtime: 2023-06-06T11:58:38.614099+02:00 -> 2023-06-07T01:48:24.724694+02:00] setup.py
> ```
> 
> * when using `type` it shows a type change `[- -> -]` even if the file type did not change, but just other mode bits.
> * `link` says `changed link`, but not how.
> * `link`, `blkdev`, `chrdev`, `fifo`, `directory` these somehow feel wrong, maybe need more thinking about practical use.

the source code of change about link, blkdev, chrdev, fifo directory  is inherit from origin:

```python
if self._item1.is_dir() or self._item2.is_dir():
	changes.append(self._presence_diff('directory'))

if self._item1.is_blk() or self._item2.is_blk():
    changes.append(self._presence_diff('blkdev'))

if self._item1.is_chr() or self._item2.is_chr():
    changes.append(self._presence_diff('chrdev'))

if self._item1.is_fifo() or self._item2.is_fifo():
    changes.append(self._presence_diff('fifo'))
...
def _presence_diff(self, item_type):
	if not self._item1.get('deleted') and self._item2.get('deleted'):
    	chg = 'removed ' + item_type
        return ({"type": chg}, chg)
    if self._item1.get('deleted') and not self._item2.get('deleted'):
        chg = 'added ' + item_type
        return ({"type": chg}, chg)
```

maybe this is a prospective error ?

**Comment by ThomasWaldmann** [comment]
"stuff item 2" can be fixed in line 1069 - there you add "iso" in front of the key.

**Comment by ThomasWaldmann** [comment]
"stuff item 5" ok, if it is a pre-existing problem, we can just say it is not in the scope of this PR to fix it.

**Comment by ThomasWaldmann** [comment]
OK, guess this PR is close to be mergable.

But when looking at the commit history in here, it is quite long and rather messy.

Would you like to clean up there yourself or shall I just squash all together into 1 commit when merging it?

**Comment by RF-Tar-Railt** [comment]
> OK, guess this PR is close to be mergable.
> 
> 
> 
> But when looking at the commit history in here, it is quite long and rather messy.
> 
> 
> 
> Would you like to (carefully!) clean up there yourself or shall I just squash all together into 1 commit when merging it?

squash better ( I have tried to rebase -i before but has caused a lot of conflict )

**Comment by ThomasWaldmann** [comment]
ok, can you fix 1069?

**Review by ThomasWaldmann — APPROVED** [review]
LGTM!

**Comment by ThomasWaldmann** [comment]
@RF-Tar-Railt guess we rather want an issue on the issue tracker.

**Comment by ThomasWaldmann** [comment]
Merged! Thanks for patiently working on this.

**Comment by RF-Tar-Railt** [comment]
can close #4634
